designs/nodeclaim-termination-grace-period.md

```diff
@@ -0,0 +1,168 @@
+
+Consider the following scenario.
+1. A Deployment has 3 replicas, each with a terminationGracePeriodSeconds of 10m.
+2. Each of these pods requires the full 10m to terminate gracefully.
+3. The Deployment has an associated PDB with maxUnavailable 1 requiring that only one of these pods be disrupted at a time.
+
+1. The cluster admin has configured NodePool.spec.disruption.terminationGracePeriod at 15m.
+2. A full node reconcilliation has started which has caused all nodes which contain the pods above to enter a deleting state at the same time.
+
+During the standard node deletion process, k8s would:
+1. Evict pod A only, which takes 10m of the available 15m to terminate gracefully. Pod B and C are not evicted, due to their PDB only allowing 1 disruption at a time.
+2. After pod A starts up healthy on a new node, pod B is evicted, but has only 5m of the needed 10m to terminate. Both pods B and C are evicted before graceful termination could complete, and pod C never received a termination signal.
+
+In order to improve this case for users of the cluster, we can assume that a pod's terminationGracePeriodSeconds is a good estimate of the time it takes to terminate gracefully and start the eviction process early if the pod would not be able to terminate within its terminationGracePeriodSeconds when a node timeout is looming.
+Pods that are deleted in this way should have a clear event log of what deleted their pod (karpenter) and why (current time > node terminationGracePeriod - pod terminationGracePeriod).
+
+Risks
+1. This behavior could be confusing to cluster administrators, and could result in immediate pod evictions when a NodePool terminationGracePeriod is set, even to a value of hours.
+2. This behavior could be confusing to users, who may have set their pod terminationGracePeriodSeconds to higher than necessary values and have their pods suddenly deleted when a node begins deletion.
+
+### Documented Warnings
+This feature carries some risk for cluster administrators given that it can bypass the configured grace periods and PDBs set by users of a cluster. These risks should be called out explicitly in the docs for the feature to minimize the risk of a cluster administrator unintentionally disrupting their users' workloads.
+1. Enabling this feature by setting NodePool.spec.template.terminationGracePeriod to any value will result in a drift reconciliation that will replace all nodes. The new terminationGracePeriod value will not be active until new NodeClaims have been provisioned.
+2. Enabling this feature can result in termination of user pods in violation of their configured grace periods or PDBs.
+3. Because of the optimization stated above, user pods could be terminated in violdation of grace periods and PDBs **before** the configured NodeClaim grace period is reached.
```

Comment:
This shouldn't be in violation of grace periods right, just PDBs? Grace periods would only be disrespected when the TGP has expired.

Reply:
Grace periods could be violated if the pod's TGPS is greater than the NodeClaim's TGP. I'll reword this.